src/Gateway.php

```diff
@@ -0,0 +1,305 @@
+        ]);
+    }
+
+    private function scheduleReconnect()
+    {
+        $this->scheduledReconnect = $this->loop->addTimer(5, function () {
+            $this->reconnect(true, true);
+        });
+    }
+
+    private function cancelScheduledReconnect()
+    {
+        if (!isset($this->scheduledReconnect)) {
+            return;
+        }
+
+        $this->loop->cancelTimer($this->scheduledReconnect);
+
+        $this->scheduledReconnect = null;
+    }
+
+    private function identify()
+    {
+        $this->sendPayload([
+            'op' => 2,
+            'd' => [
+                'token' => $this->token,
+                'intents' => $this->intents->get(),
+                'properties' => [
+                    'os' => PHP_OS,
+                    'browser' => 'Exan\Fenrir',
+                    'device' => 'Exan\Fenrir',
+                ]
+            ]
+        ]);
+
+        $this->shouldIdentify = false;
+    }
+
+    private function handlePayload(Payload $payload)
+    {
+        switch ($payload->op) {
+            /**
+             * "Regular" events
+             */
+            case 0:
+                if (isset($payload->s) && $payload->s > $this->sequence) {
+                    $this->sequence = $payload->s;
+                }
+
+                $this->handleEvent($payload);
+                break;
+
+            /**
+             * Resume event
+             */
+            case 7:
+                $this->reconnect(true, true);
+                break;
+
+            /**
+             * Invalid session
+             */
+            case 9:
+                $this->reconnect(
+                    false,
+                    isset($payload->d) && $payload->d === true
+                );
+                break;
+
+            /**
+             * Hello event
+             */
+            case 10:
+                if ($this->shouldIdentify) {
+                    $this->identify();
+                } else {
+                    $this->resume();
+                }
+
+                $this->handleHello($this->mapper->map($payload->d, new Hello()));
+                break;
+
+            /**
+             * Acknowledgement of heartbeat
+             */
+            case 11:
+                $this->cancelScheduledReconnect();
+                break;
+        }
+    }
+
+    private function handleHello(Hello $data)
+    {
+        $this->heartbeatInterval = $data->heartbeat_interval;
+
+        $this->startHeartbeat();
+    }
+
+    private function handleEvent(Payload $payload)
+    {
+        if ($payload->t === Events::READY && isset($payload->d->session_id, $payload->d->resume_gateway_url)) {
+            $this->sessionId = $payload->d->session_id;
+            $this->reconnectUrl = $payload->d->resume_gateway_url;
+        }
+
+        $this->events->handle($payload);
+    }
+
+    private function sendPayload(array $data, bool $useBucket = true)
+    {
+        $this->websocket->send(json_encode($data), $useBucket);
```

Comment:
Would it be possible to have `$this->websocket->sendJson(string|array|jsonSerializable $json)` that json_encodes internally?

Reply:
https://github.com/dc-Ragnarok/Fenrir/issues/3

Out of scope for this PR